client/api/README.md

```diff
@@ -0,0 +1,257 @@
+
+Internally, there is only one internal server `struct` per API group, that handles all the versions for that API group. That server is defined in this repo's `internal/server/<api_group_name>` (e.g. `internal/server/iscsi`) go package. This go package should follow the following pattern:
+
+<a name="serverPkgTree"></a>
+```
+internal/server/<api_group_name>
+├── internal
+│   └── types.go
+└── server.go
+```
+where `types.go` should contain the internal types corresponding to the various protobuf types for that API group - these internal structures must be able to represent all the different versions of the API. For example, given a `dummy` API group with two versions defined by the following `proto` files:
+
+`client/api/dummy/v1alpha1/api.proto`
+```proto
+syntax = "proto3";
+
+package v1alpha1;
+
+service Dummy {
+    // ComputeDouble computes the double of the input. Real smart stuff!
+    rpc ComputeDouble(ComputeDoubleRequest) returns (ComputeDoubleResponse) {}
+}
+
+message ComputeDoubleRequest{
+    int32 input32 = 1;
+}
+
+message ComputeDoubleResponse{
+    int32 response32 = 1;
+}
+```
+
+and
+
+`client/api/dummy/v1/api.proto`
+```proto
+syntax = "proto3";
+
+package v1;
+
+service Dummy {
+    // ComputeDouble computes the double of the input. Real smart stuff!
+    rpc ComputeDouble(ComputeDoubleRequest) returns (ComputeDoubleResponse) {}
+}
+
+message ComputeDoubleRequest{
+    // we changed in favor of an int64 field here
+    int64 input = 2;
+}
+
+message ComputeDoubleResponse{
+    int64 response = 2;
+    
+    // set to true if the result overflowed
+    bool overflow = 3;
+}
+```
+
+then `internal/server/dummy/internal/types.go` could look something like:
+```go
+type ComputeDoubleRequest struct {
+	Input int64
+}
+
+type ComputeDoubleResponse struct {
+	Response int64
+	Overflow bool
+}
+```
+and then the API group's server (`internal/server/dummy/server.go`) needs to define the callbacks to handle requests for all API versions, e.g.:
+```go
+type Server struct{}
+
+func (s *Server) ComputeDouble(ctx context.Context, request *internal.ComputeDoubleRequest, version apiversion.Version) (*internal.ComputeDoubleResponse, error) {
+	in := request.Input64
+	out := 2 * in
+	
+	response := &internal.ComputeDoubleResponse{}
+
+	if sign(in) != sign(out) {
+		// overflow
+		response.Overflow = true
+	} else {
+		response.Response = out
+	}
+
+	return response, nil
+}
+
+func sign(x int64) int {
+	switch {
+	case x > 0:
+		return 1
+	case x < 0:
+		return -1
+	default:
+		return 0
+	}
+}
+```
+All the boilerplate code to:
+ * add a named pipe to the server for each version of the API group, listening for its version's requests, and replying with its version's responses
```

Comment:
how about: "listening for its version's requests" => "listen for each version's request"
and: "replying with it's version's responses" => "reply with each version's responses"

Reply:
👍